### PR TITLE
Removed state for removed accounts

### PR DESCRIPTION
**Pull request by hsato03**
### Description

When deleting an account, its state is maintained as `Enabled` even if it is removed.

Accordingly the `Removed` state for accounts was created, which is always set when an account is deleted.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [X] Minor
- [ ] Trivial

### How Has This Been Tested?

I created and deleted an account. Then, I created a SELECT SQL query for the `account` table and the account state was `Removed`.

### CODE REVIEW

**Review by DaanHoogland — APPROVED** [review]
Good change! one remark, other entities have a state DESTROYED. maybe it would make sense to be in line with that?

**Comment by weizhouapache** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@hsato03 do you want this in 4.18.1?

**Comment by hsato03** [comment]
> 



> Good change! one remark, other entities have a state DESTROYED. maybe it would make sense to be in line with that?

@DaanHoogland Yes, that's a good idea. But some entities have the `Inactive` removed state like `domain` and `vm_template`.

**Review by soreana — APPROVED** [review]

**Comment by weizhouapache** [comment]
> Good change! one remark, other entities have a state DESTROYED. maybe it would make sense to be in line with that?

@DaanHoogland 
to be honest, I think `removed` makes more sense.

**Comment by DaanHoogland** [comment]
@weizhouapache REMOVED does make sense but having different terms with the same meaning is something we should prevent, and adding a new one should replace all occurences of the old ones. Not going to block this for that reason, but it will be confusing for new operators and developers alike.

worth a discuss thread?

cc @soreana @hsato03

**Comment by weizhouapache** [comment]
@hsato03 
since this is not a bug, but an improvement. let's move to next major release 4.19.0.0
cc @DaanHoogland

**Comment by DaanHoogland** [comment]
I think this can be included in 4.18.1 as well, but do not mind either way

File: api/src/main/java/com/cloud/user/Account.java
Comment (rohityadavcloud) [inline api/src/main/java/com/cloud/user/Account.java:33]: Can we call it REMOVED ?

**Review by rohityadavcloud — APPROVED** [review]
LGTM - I'm not sure if there could be only possible side-effects; or it should also emit some kind of event for usage-server?

**Comment by weizhouapache** [comment]
> LGTM - I'm not sure if there could be only possible side-effects; or it should also emit some kind of event for usage-server?

it is one of my concerns as well.
let's consider it for 4.19.0.0

**Comment by soreana** [comment]
@blueorangutan package

**Comment by hsato03** [comment]
I agree with @rohityadavcloud and @weizhouapache, REMOVED sounds better for accounts. @DaanHoogland Although your idea of ​​standardizing the state of entities is good, there are already other entities that use the REMOVED state.

> LGTM - I'm not sure if there could be only possible side-effects; or it should also emit some kind of event for usage-server?

@rohityadavcloud ACS already emits an event when deleting an account.

**Comment by soreana** [comment]
@blueorangutan package

**Comment by soreana** [comment]
@DaanHoogland @rohityadavcloud @weizhouapache I did some research to understand the definitions of `removed` and `destroyed` in CloudStack terminology.

Once an object is Destroyed in CloudStack, it remains visible to the root admin, who has the exclusive ability to either restore or remove it. Even when an object is destroyed, it can still be restored. But once an object is removed, it cannot be restored and will be lost permanently.

As of now, it is not possible to restore the account that was removed. Therefore, using the term `Remove` state would be more appropriate than `Destroyed`.

**Comment by DaanHoogland** [comment]
have you tested @soreana ?

**Comment by hsato03** [comment]
Hi @soreana,

Any updates on this one?

**Comment by soreana** [comment]
Hi @DaanHoogland @hsato03,

I've tested the upgrade from 4.18.1 to 4.19 the process went smooth and the account have removed state now.

Before upgrade:
```
MariaDB [cloud]> select account_name,state,removed from account ;
+--------------------------+---------+---------------------+
| account_name             | state   | removed             |
+--------------------------+---------+---------------------+
| system                   | enabled | NULL                |
| admin                    | enabled | NULL                |
| baremetal-system-account | enabled | NULL                |
| Removed-01               | enabled | 2023-09-28 08:41:35 |
| Removed-02               | enabled | 2023-09-28 08:45:18 |
| Removed-03               | enabled | 2023-09-28 08:46:16 |
+--------------------------+---------+---------------------+
6 rows in set (0.001 sec)

MariaDB [cloud]>
```

After upgrade:
```
MariaDB [cloud]> select account_name,state,removed from account ;
+--------------------------+---------+---------------------+
| account_name             | state   | removed             |
+--------------------------+---------+---------------------+
| system                   | enabled | NULL                |
| admin                    | enabled | NULL                |
| baremetal-system-account | enabled | NULL                |
| Removed-01               | removed | 2023-09-28 08:41:35 |
| Removed-02               | removed | 2023-09-28 08:45:18 |
| Removed-03               | removed | 2023-09-28 08:46:16 |
+--------------------------+---------+---------------------+
6 rows in set (0.001 sec)

MariaDB [cloud]>
```

**Comment by hsato03** [comment]
@soreana Thank's for testing.

**Pull request merged**
